### DISCUSSION TITLE
custom species additions to 3k model?

### DISCUSSION THREAD

**Discussion by dripp33** [mcguirepr89/BirdNET-Pi#841] [Ideas]
I much prefer the 3k model (in the US) in terms of speed, accuracy, and in reducing the number of exclusions needed, but would be great if it would be possible to add certain species to the location-modified species list. I'm finding that certain nocturnally active species, particularly some species of owls that are found in my area, don't show up in my species list unless I set the species occurrence frequency threshold so low as to be basically not very useful. Would be great if one could add specific species to the location-modified species list. I tend to think this is not as simple as adding a species label to the labels list because it seems like the labels list would need to be in sync with the Tflite data set (although I would love to be wrong about this).

Thanks for the great software and any info on if it's somehow currently possible to manually add certain species to the 3k model for a particular area and frequency threshold.
--dave

**Comment by ehpersonal38** [comment]
This is possible! If you edit `~/BirdNET-Pi/scripts/server.py` in a text editor, you'll see a variable called `PREDICTED_SPECIES_LIST` on line 39. You can add your desired species to this array, using the same species format you see in `~/BirdNET-Pi/model/labels.txt`. Example:

Line 39, before:
`PREDICTED_SPECIES_LIST = []`

After:
`PREDICTED_SPECIES_LIST = ["Cardinalis Cardinalis_Northern Cardinal", "Accipiter bicolor_Bicolored Hawk"]`

Then you'll be able to save this file and the model will now include these species in the list of species it will try to detect. The caveat is that when you go to update your BirdNET-Pi these changes will be overwritten, so I'd advise you add `scripts/server.py` to your `~/BirdNET-Pi/.gitignore` file.

**Reply by dripp33** [reply]
That's great, thank you for the info! I was hoping there was the capability to do this, just couldn't find it. Would be nice if this was in the documentation somewhere (perhaps I just missed it), as folks using the 3k model may be missing detections of some nocturally active species that get underrepresented in reported sightings lists.

Thanks again!
--dave

**Reply by dripp33** [reply]
I tried this out and when I add species (i.e., 3 types of owls) to the PREDICTED_SPECIES_LIST set in /scripts.server.py the analyzer software ONLY finds the species that I added to the set (rather than adding these species to my 3k model location-generated set). I wanted to add a few nocturnal species (i.e., owls) that are relatively common in my area, but are not within my 3k model location occurrence threshold. Hence the software ONLY detected a couple of owls rather than the 1000 or so detections that it normally finds over 24 hours.

Please let me know if works differently for you.  Thanks,
--dave

**Reply by ehpersonal38** [reply]
Oops, you're totally right! Sorry for that oversight.

I think to fix this you'll want to add your `PREDICTED_SPECIES_LIST = [{your species here}]` to line 158. See here:
![sublime_text_eCxIWX37PX](https://user-images.githubusercontent.com/103586016/231635033-abf59962-e961-4c0a-aef7-28cc06f24cc8.png)


^make sure you also add `global PREDICTED_SPECIES_LIST` like you see.

Then be sure to set `PREDICTED_SPECIES_LIST` back to an empty list on line 40. Restart your services and you should be good to go.

**Reply by dripp33** [reply]
Thanks! I'll give that a shot.
--dave

**Reply by dripp33** [reply]
Seems to work! Thanks again.
--dave

**Comment by Swiftnesses** [comment]
Sorry to bring this back from the dead, I'm also trying to add some custom owls, is this correct?

```
def predictSpeciesList(lat, lon, week):
    global PREDICTED_SPECIES_LIST
    PREDICTED_SPECIES_LIST = ["Tyto alba_Barn Owl","Strix aluco_Tawny Owl","Asio flammeus_Short-eared Owl","Athene noctua_Little Owl"]

    l_filter = explore(lat, lon, week)
    for s in l_filter:
        if s[0] >= float(sf_thresh):
            # if there's a custom user-made include list, we only want to use the species in that
            if (len(INCLUDE_LIST) == 0):
                PREDICTED_SPECIES_LIST.append(s[1])
```
